### ISSUE TITLE
Questions about baseline implementations

### ISSUE THREAD

**Issue by blue-531** [TilemahosAravanis/Retrieve-and-Segment#3]
Thanks for releasing the code and for your interesting work!

We are building on RnS and would like to reproduce your comparisons as faithfully as possible, rather than relying on our own reimplementations. We had two questions about the experimental setup:

1. **kNN-CLIP** The paper mentions that you “follow the official implementation,” but we could not find a public release of the code. Could you point us to the official implementation? If it was shared privately, we would also appreciate knowing whom we could contact for access.
2. Would you consider releasing your implementation used for both kNN-CLIP and FreeDA? Even an unpolished script or a minimal implementation would be very helpful for ensuring a fair and faithful comparison.

Thanks in advance for your help, and thanks again for making your work available!

**Comment by TilemahosAravanis** [comment]
Hi @blue-531, thanks for the kind words and for the careful reading!

**1. kNN-CLIP "official implementation"**

To our knowledge the kNN-CLIP authors have not released code (there is no code link on the arXiv / OpenReview pages, and we did not obtain code privately either). The sentence in the supplementary (Sec. 6.3, "we follow the official implementation") is imprecise and we will correct it; the main paper (Sec. 4.1) states it correctly: **we re-implemented kNN-CLIP** from the description in their paper (Sec. 3.1-3.2, Eqs. 1-4) inside our own pipeline. For the original code, the kNN-CLIP authors would be the ones to contact.

**2. Our implementations of kNN-CLIP and FreeDA**

Both competitors are plugged into exactly the same pipeline as RNS, so they use the same backbone features (OpenCLIP ViT-B/16 with the MaskCLIP trick, or DINOv3.txt ViT-L/16), the same visual support set (the per-image, GT-mask-pooled visual class features that form the RNS support bank, `support_vlm_feats` in `main_eval.py`), the same textual class features, and the same prediction granularity (patches, or SAM 2.1 regions). Only the classification rule differs:

*kNN-CLIP*
- **Database:** the support bank (one feature per support image and class, labels known). It replaces the DINOv2 mask-pooled database of the original paper.
- **Query:** the test patch (or SAM-region) features. k-NN search by cosine similarity with FAISS, k = 4 (the same k as RNS).
- **Retrieval prediction:** the k similarities are softmax-normalised and accumulated per class through the neighbour labels (confidence-aware pseudo-logits, Eq. 3 of kNN-CLIP).
- **Fusion with zero-shot (Eq. 4):** if the maximum zero-shot probability (softmax of 100 x cosine to the text features) is >= T, the zero-shot prediction is kept, otherwise the retrieval prediction is used (lambda = 1). T = 0.9 for all datasets and all shots.

*FreeDA*
- We ported the inference rule of the official code (default config) to real visual support: one prototype per class = mean of that class's per-image support features. The diffusion-generated collection and the text-to-prototype retrieval are dropped, since the labels of the support features are known.
- **Local similarities:** sigmoid of the cosine between each patch / region feature and the class prototypes (their region-pooled backbone similarities, with SAM 2.1 regions instead of Felzenszwalb superpixels).
- **Global similarities:** cosine between the global image feature (mean of the test image's patch / region features) and the text features. This replaces the separate CLIP image embedding, so that a single backbone is used.
- **Ensemble:** beta * local + (1 - beta) * global with beta = 0.8 (their `ensemble_dino_clip`), then argmax over classes.

**Code.** Below is the exact logic, written against the released code base (drop-in next to `core/rns.py`; it reuses `retrieve_similar_feats`, `support_vlm_feats`, `textual_embeddings`, `k`, `use_text` and `drop_classes` exactly as `RNS_Inference` does). We will push it to the repo together with a `method.name` switch (`RNS` / `KNN-CLIP` / `FREEDA`), so that both baselines run with the same commands, configs and seeds as RNS.

```python
# core/baselines.py -- kNN-CLIP and FreeDA on the RNS support bank.
# Same conventions as RNS_Inference in core/rns.py: same backbone features, same support
# set (vlm.support_vlm_feats) and same patch / SAM-region granularity.
import torch
import torch.nn.functional as F
from core.rns import retrieve_similar_feats


def _query_feats(vlm_features):
    """(1, C, N) dense or region features -> L2-normalised (N, C), as in RNS_Inference."""
    B, C = vlm_features.shape[:2]
    x = vlm_features.reshape(B, C, -1)
    x = x / x.norm(dim=1, keepdim=True)
    return x.permute(0, 2, 1).squeeze(0)


def kNN_CLIP_Inference(vlm, vlm_features, T=0.9):
    """kNN-CLIP (Gui et al., TMLR 2024) re-implemented on top of the RNS support bank.

    database : vlm.support_vlm_feats -- one GT-mask-pooled feature per (support image, class),
               label = row index % num_classes. Rows of classes without visual support are +inf
               and are never returned by the inner-product search.
    query    : test patch features, or SAM-region features when mask_proposal_strategy="SAM".
    T        : confidence threshold on the zero-shot prediction (Eq. 4 of kNN-CLIP, lambda = 1).
               T = 1.0 (or use_text = false) gives the 'w/o text' variant.
    """
    q = _query_feats(vlm_features)                                               # (N, C)

    sims, idx = retrieve_similar_feats(vlm, q, vlm.support_vlm_feats, vlm.k)     # (N, k)
    sims = sims.masked_fill(idx < 0, float('-inf'))                              # fewer than k hits
    labels = F.one_hot((idx.clamp(min=0) % vlm.num_classes).long(), vlm.num_classes).float()

    # confidence-aware pseudo-logits: neighbour similarities softmax-normalised over the
    # k neighbours, then accumulated per class. (Eq. 3 of kNN-CLIP applies the softmax to the
    # per-class sums instead; both are monotone in the class-wise similarity evidence.)
    P_ret = torch.einsum('nk,nkc->nc', F.softmax(sims, dim=-1), labels)         # (N, C)

    if not vlm.use_text or T >= 1.0:
        return P_ret

    P_base = F.softmax((q @ vlm.textual_embeddings.T) * 100, dim=1)             # zero-shot (N, C)
    confident = (P_base.max(dim=-1).values >= T).unsqueeze(1)
    return torch.where(confident, P_base, P_ret)


def FreeDA_Inference(vlm, vlm_features, beta=0.8):
    """FreeDA (Barsellotti et al., CVPR 2024) with real visual support.

    The diffusion-generated prototype collection is replaced by the RNS support bank and the
    text->prototype retrieval is skipped because support labels are known:
    prototype_c = normalised mean of the per-image visual class features of class c
    (FreeDA's `prototype_mean`). Local patch / region similarities to the prototypes are
    ensembled with the global image-to-text similarities with weight beta
    (FreeDA's `ensemble_dino_clip` = 0.8). beta = 1.0 (or use_text = false) gives 'w/o text'.
    """
    q = _query_feats(vlm_features)                                               # (N, C)

    bank = vlm.support_vlm_feats.clone()
    bank[bank == float('+inf')] = 0.0
    protos = bank.reshape(-1, vlm.num_classes, vlm.vlm_feat_dim).mean(dim=0)
    protos = protos / protos.norm(dim=1, keepdim=True)
    if len(vlm.drop_classes) > 0:   # classes w/o visual support: mean of the available prototypes
        keep = torch.ones(vlm.num_classes, dtype=torch.bool, device=protos.device)
        keep[vlm.drop_classes] = False
        protos[vlm.drop_classes] = protos[keep].mean(dim=0, keepdim=True)
        protos = protos / protos.norm(dim=1, keepdim=True)

    local = torch.sigmoid(q @ protos.T)                                          # (N, C)

    g = q.mean(dim=0)
    g = g / g.norm()
    glob = g @ vlm.textual_embeddings.T                                          # (C,)

    if not vlm.use_text:
        beta = 1.0
    return F.softmax(beta * local + (1 - beta) * glob.unsqueeze(0), dim=1)
```

Wiring: in the model's `forward`, return `kNN_CLIP_Inference(self, vlm_features)` or `FreeDA_Inference(self, vlm_features)` instead of `RNS_Inference(self, vlm_features)` depending on the method; region-to-pixel mapping and evaluation are unchanged. All other settings (crop sizes, strides, support seeds) are the ones in the README / configs, and the paper numbers are averages over the seeds listed in the README.

Hope this helps, and let us know if anything is unclear!

**Comment by azfunder** [comment]
Question about the OVS framing vs. the few-shot setting

Dear authors,

Thank you very much for releasing the code and for this excellent work !

I do have one conceptual question about how the work is positioned. The paper is framed as open-vocabulary segmentation (OVS), but the core method relies on a support set of pixel-annotated images (i.e., a few-shot setting): visual support features are retrieved and fused with textual features to train a lightweight per-image classifier.

My understanding is that "open-vocabulary" is usually understood as zero-shot — segmenting arbitrary categories specified only by a text prompt, without ever seeing annotated examples of those categories. Adding a set of pixel-annotated support images seems closer to few-shot / in-context segmentation than to pure OVS.

Could you clarify how the method still qualifies as "open-vocabulary"? A few specific points I'd love your view on:

In the partial-visual-support setting, some classes lack visual examples — is it this ability (handling classes without visual support via text) that you consider the key "open-vocabulary" property?
Is the claim more about preserving text-based zero-shot predictions while boosting accuracy on supported classes, rather than about being purely zero-shot in all cases?
Compared to in-context segmentation (e.g., COSINE), how would you characterize the distinction?
Thank you again for your time and for sharing this work!
